File: scripts/koruh/blocks/teleunit.js
const diriumColor = Color.valueOf("96f7c3");
const diriumColor2 = Color.valueOf("ccffe4");

const tpCoolDown = new StatusEffect("tpcooldown"); //empty effect for tp cooldown
tpCoolDown.color = diriumColor2;
tpCoolDown.effect = Fx.none;

const tpOut = new Effect(30, e => {
    Draw.color(diriumColor);
    Lines.stroke(3*e.fout());
    Lines.square(e.x, e.y, e.finpow() * e.rotation, 45);
    Lines.stroke(5*e.fout());
    Lines.square(e.x, e.y, e.fin() * e.rotation, 45);
    Angles.randLenVectors(e.id, 10, e.fin() * (e.rotation + 10), (x, y) => {
        Fill.square(e.x + x, e.y + y, e.fout() * 4, 100*Mathf.randomSeed(e.id+1)*e.fin());
    });
});

const tpIn = new Effect(50, e => {
    if(!(e.data instanceof UnitType)) return;
    var region = e.data.icon(Cicon.full);
    Draw.color();
    Draw.mixcol(diriumColor, 1);
    Draw.rect(region, e.x, e.y, region.width * Draw.scl * e.fout(), region.height * Draw.scl * e.fout(), e.rotation);
    Draw.mixcol();
});

const tpFlash = new Effect(30, e => {
    if(!(e.data instanceof Unit) || e.data.dead) return;
    var region = e.data.type.icon(Cicon.full);
    Draw.mixcol(diriumColor2, 1);
    Draw.alpha(e.fout());
    Draw.rect(region, e.data.x, e.data.y, e.data.rotation - 90);
    Draw.mixcol();
    Draw.color();
});
tpFlash.layer = Layer.flyingUnit + 1;

const teleunit = extendContent(Block, "teleunit", {
    load() {
        this.super$load();
        this.lightRegion = Core.atlas.find(this.name + "-lights");
        this.topRegion = Core.atlas.find(this.name + "-top");
        this.arrowRegion = Core.atlas.find("transfer-arrow");
    },
    init(){
        this.super$init();
        this.buildId = [];
    }
});
teleunit.update = true;
teleunit.solid = false;
//teleunit.consumesTap = true;
teleunit.ambientSound = Sounds.techloop;
teleunit.ambientSoundVolume = 0.02;
teleunit.configurable = true;

//WHEN PORTING TO JAVA, PLZ MAKE IT TELEPORT PAYLOADS
//teleunit.outputsPayload = true;
//teleunit.outputFacing = false;

//const d4x = [1, 0, -1, 0], d4y = [0, 1, 0, -1];

teleunit.buildType = prov(() => extend(Building, {
    _warmup: 0,
    _warmup2: 0,
    updateTile(){
        this._warmup = Mathf.lerpDelta(this._warmup, this.consValid() ? 1 : 0, 0.05);
        this._warmup2 = Mathf.lerpDelta(this._warmup2, this.consValid() && this.enabled ? 1 : 0, 0.05);
    },
    draw() {
        this.super$draw();
        Draw.color(Color.white);
        Draw.alpha(0.45 + Mathf.absin(Time.time, 7, 0.26));
        Draw.rect(teleunit.topRegion, this.x, this.y);
        if(this._warmup >= 0.001){
            Draw.z(Layer.bullet);
            Draw.color(diriumColor, this.team.color, Mathf.absin(Time.time, 19, 1));
            Lines.stroke((Mathf.absin(Time.time, 62, 0.5) + 0.5) * this._warmup);
            Lines.square(this.x, this.y, 10.5, 45);
            if(this._warmup2 >= 0.001){
                Lines.stroke((Mathf.absin(Time.time, 62, 1) + 1) * this._warmup2);
                Lines.square(this.x, this.y, 8.5, Time.time / 2);
                Lines.square(this.x, this.y, 8.5, -1 * Time.time / 2);
            }
        }
        Draw.reset();
    },
    drawSelect(){
        Draw.color(this.consValid() ? (this.inRange(Vars.player) ? diriumColor : Pal.accent) : Pal.darkMetal);
        var length = Vars.tilesize * teleunit.size / 2 + 3 + Mathf.absin(Time.time, 5, 2);

        Draw.rect(teleunit.arrowRegion, this.x + length, this.y, (0 + 2) * 90);
        Draw.rect(teleunit.arrowRegion, this.x, this.y + length, (1 + 2) * 90);
        Draw.rect(teleunit.arrowRegion, this.x + -1 * length, this.y, (2 + 2) * 90);
        Draw.rect(teleunit.arrowRegion, this.x, this.y + -1 * length, (3 + 2) * 90);

        Draw.color();
    },
    shouldAmbientSound(){
        return this.consValid();
    },
    created(){
        this.super$created();
        teleunit.buildId[this.id] = this;
    },
    onRemoved(){
        teleunit.buildId[this.id] = null;
        this.super$onRemoved();
    },
    getDestList(){
        var str = this.power.graph.toString();

        str = str.substring(str.indexOf("consumers={")+11,str.indexOf("batteries")-3);
        var arr = str.split(", ");
        //print("Arr: "+arr);
        //print("BId: "+teleunit.buildId);
        var barr = [];
        for(var i=0; i<arr.length;i++){
            var tid = arr[i].split("#")[1];
            //print("tid: "+tid);
            if(teleunit.buildId[tid] == undefined || teleunit.buildId[tid] == null || !teleunit.buildId[tid] || (!teleunit.buildId[tid].enabled && teleunit.buildId[tid] != this)) continue;
            barr.push(teleunit.buildId[tid]);
        }
        barr.sort(function(a, b) {
            return a.pos() - b.pos();
        });
        return barr;
    },
    inRange(player){
        return player.unit() != null && !player.unit().dead && Math.abs(player.unit().x - this.x) <= 2.5 * Vars.tilesize && Math.abs(player.unit().y - this.y) <= 2.5 * Vars.tilesize;
    },
    shouldShowConfigure(player){
        return this.consValid() && this.inRange(Vars.player);
    },
    configTapped(){
        if(!this.consValid() || !this.inRange(Vars.player)) return false;
        this.configure(null);
        Sounds.click.at(this);
        return false;
    },
    configured(unit, value){
        if(unit != null && unit.isPlayer() && !(unit instanceof BlockUnitc)) this.tpPlayer(unit.getPlayer());
    },
    tpPlayer(player){
        this.tpUnit(player.unit(), player == Vars.player);
        if(Vars.player != null && player == Vars.player) Core.camera.position.set(player);
    },
    tpUnit(unit, isPlayer){
        var barr = this.getDestList();
        if(barr.length <= 0) return;
        var index = barr.indexOf(this);
        if(index < 0){
            print("Error! Origin pad not in list!");
        }
        index++;
        if(index >= barr.length) index = 0;
        var dest = barr[index];
        if(!Vars.headless) tpIn.at(unit.x, unit.y, unit.rotation - 90, Color.white, unit.type);
        unit.set(dest.x, dest.y);
        unit.snapInterpolation();
        unit.set(dest.x, dest.y);//for good measure

        if(!Vars.headless) this.effects(dest, unit.hitSize * 1.7, isPlayer, unit);
    },
    effects(dest, hitSize, isPlayer, unit){
        //TODO: EoD-style total unit effect
        if(isPlayer){
            Sounds.plasmadrop.at(dest.x, dest.y, Mathf.random() * 0.2 + 1);
            Sounds.lasercharge2.at(this.x, this.y, Mathf.random() * 0.2 + 0.7);
        }
        else{
            Sounds.plasmadrop.at(this.x, this.y, Mathf.random() * 0.2 + 1);
            Sounds.lasercharge2.at(dest.x, dest.y, Mathf.random() * 0.2 + 0.7);
        }
        tpOut.at(dest.x, dest.y, hitSize);
        tpFlash.at(dest.x, dest.y, 0, Color.white, unit);
    },
    unitOn(unit){
        if(!this.consValid()) return;
        if(unit.hasEffect(tpCoolDown) || unit.isPlayer()) return;
        this.tpUnit(unit, false);
        unit.apply(tpCoolDown, 120);
    },
    consValid(){
        //suppress considering enabled
        return this.power.status > 0.98;
    }
}));

File: main/src/unity/world/blocks/units/TeleUnit.java
package unity.world.blocks.units;

import arc.*;
import arc.graphics.*;
import arc.graphics.g2d.*;
import arc.math.*;
import arc.math.geom.*;
import arc.util.*;
import mindustry.game.EventType.*;
import mindustry.game.*;
import mindustry.gen.*;
import mindustry.graphics.*;
import mindustry.world.*;
import mindustry.world.blocks.payloads.*;
import unity.content.*;
import unity.graphics.*;

import static arc.Core.*;
import static mindustry.Vars.*;

public class TeleUnit extends Block{
    //this class was made assuming only one instance. if more, static should be removed or further improvement.
    protected static final TeleUnitBuild[] heads = new TeleUnitBuild[Team.baseTeams.length];
    protected static final int[] listSizes = new int[Team.baseTeams.length];
    protected TextureRegion lightRegion, topRegion, arrowRegion;

    public TeleUnit(String name){
        super(name);
        update = configurable = outputsPayload = true;
        Events.on(WorldLoadEvent.class, e -> {
            for(int i = 0; i < heads.length; i++){
                heads[i] = null;
                listSizes[i] = 0;
            }
        });
    }

    @Override
    public void load(){
        super.load();
        lightRegion = atlas.find(name + "-lights");
        topRegion = atlas.find(name + "-top");
        arrowRegion = atlas.find("transfer-arrow");
    }

    public class TeleUnitBuild extends Building{
        protected float warmup, warmup2;
        //slow adding O(n), but fast removing O(1).
        //I hope non-primitive fields are just pointer variable.
        protected TeleUnitBuild next, prev;
        protected Team previousTeam;

        protected boolean isTeamChanged(){
            return previousTeam != team;
        }

        @Override
        public void updateTile(){
            warmup = Mathf.lerpDelta(warmup, canConsume() ? 1f : 0f, 0.05f);
            warmup2 = Mathf.lerpDelta(warmup2, canConsume() && enabled ? 1f : 0f, 0.05f);
            if(isTeamChanged()){
                onRemoved();
                created();
            }
        }

        @Override
        public void draw(){
            super.draw();
            Draw.color(Color.white);
            Draw.alpha(0.45f + Mathf.absin(7f, 0.26f));
            Draw.rect(topRegion, x, y);
            if(warmup >= 0.001f){
                Draw.z(Layer.bullet);
                Draw.color(UnityPal.dirium, team.color, Mathf.absin(19f, 1f));
                Lines.stroke((Mathf.absin(62f, 0.5f) + 0.5f) * warmup);
                Lines.square(x, y, 10.5f, 45f);
                if(warmup2 >= 0.001f){
                    Lines.stroke((Mathf.absin(62f, 1f) + 1f) * warmup2);
                    Lines.square(x, y, 8.5f, Time.time / 2f);
                    Lines.square(x, y, 8.5f, -1 * Time.time / 2f);
                }
            }
            Draw.reset();
        }

        @Override
        public void drawSelect(){
            Draw.color(canConsume() ? inRange(player) ? UnityPal.dirium : Pal.accent : Pal.darkMetal);
            float length = tilesize * size / 2f + 3f + Mathf.absin(5f, 2f);
            Draw.rect(arrowRegion, x + length, y, 180f);
            Draw.rect(arrowRegion, x, y + length, 270f);
            Draw.rect(arrowRegion, x - length, y, 0f);
            Draw.rect(arrowRegion, x, y - length, 90f);
            Draw.color();
        }

        @Override
        public boolean shouldAmbientSound(){
            return canConsume();
        }

        @Override
        public void created(){
            previousTeam = team;
            TeleUnitBuild temp = heads[team.id];
            listSizes[team.id]++;
            if(temp == null){
                heads[team.id] = next = prev = this;
                return;
            }
            for(int i = 0, len = listSizes[team.id] - 1; i < len; i++){
                if(temp.pos() > pos()){
                    if(i == 0) heads[team.id] = this;
                    next = temp;
                    prev = temp.prev;
                    prev.next = temp.prev = this;
                    return;
                }else temp = temp.next;
            }
            next = temp;
            prev = temp.prev;
            prev.next = temp.prev = this;
        }

        @Override
        public void onRemoved(){
            int a = isTeamChanged() ? previousTeam.id : team.id;
            if(heads[a] != null){
                if(this == heads[a]){
                    if(listSizes[a] > 1) heads[a] = next;
                    else heads[a] = null;
                }
                prev.next = next;
                next.prev = prev;
                next = prev = null;
                listSizes[a]--;
            }
        }

        protected TeleUnitBuild getDest(){
            TeleUnitBuild temp = this;
            for(int i = 0, len = listSizes[previousTeam.id]; i < len; i++){
                temp = temp.next;
                if(temp != null && temp.enabled && temp.power.graph == power.graph) return temp;
            }
            return temp;
        }

        protected boolean inRange(Player player){
            return player.unit() != null && player.unit().isValid() && Math.abs(player.unit().x - x) <= 2.5f * tilesize && Math.abs(player.unit().y - y) <= 2.5f * tilesize;
        }

        @Override
        public boolean shouldShowConfigure(Player player){
            return canConsume() && inRange(player);
        }

        @Override
        public boolean configTapped(){
            if(!canConsume() || !inRange(player)) return false;
            configure(null);
            Sounds.click.at(this);
            return false;
        }

        @Override
        public void configured(Unit unit, Object value){
            if(unit != null && unit.isPlayer() && !(unit instanceof BlockUnitc)) tpPlayer(unit.getPlayer());
        }

        protected void tpPlayer(Player player1){
            tpUnit(player1.unit(), player1 == player);
            if(player != null && player1 == player) camera.position.set(player1);
        }

        protected void tpUnit(Unit unit, boolean isPlayer){
            TeleUnitBuild dest = getDest();
            if(dest == null) return;
            if(!headless) UnityFx.tpIn.at(unit.x, unit.y, unit.rotation - 90f, Color.white, unit.type);
            unit.set(dest.x, dest.y);
            unit.snapInterpolation();
            unit.set(dest.x, dest.y);
            if(!headless) effects(dest, unit.hitSize * 1.7f, isPlayer, unit);
        }

        protected void effects(TeleUnitBuild dest, float hitSize, boolean isPlayer, Unit unit){
            if(isPlayer){
                Sounds.plasmadrop.at(dest, Mathf.random() * 0.2f + 1f);
                Sounds.lasercharge2.at(this, Mathf.random() * 0.2f + 0.7f);
            }else{
                Sounds.plasmadrop.at(this, Mathf.random() * 0.2f + 1f);
                Sounds.lasercharge2.at(dest, Mathf.random() * 0.2f + 0.7f);
            }
            UnityFx.tpOut.at(dest, hitSize);
            UnityFx.tpFlash.at(dest.x, dest.y, 0f, Color.white, unit);
        }

        @Override
        public void unitOn(Unit unit){
            if(!canConsume()) return;
            if(unit.hasEffect(UnityStatusEffects.tpCoolDown) || unit.isPlayer()) return;
            tpUnit(unit, false);
            unit.apply(UnityStatusEffects.tpCoolDown, 120f);
        }

        @Override
        public boolean canConsume(){
            return power.status > 0.98f;
        }

        @Override
        public boolean acceptPayload(Building source, Payload payload){
            TeleUnitBuild dest = getDest();
            if(!canConsume() || !dest.enabled) return false;
            int ntrns = 1 + size / 2;
            Building nextBuild = dest.nearby(Geometry.d4x(source.rotation) * ntrns, Geometry.d4y(source.rotation) * ntrns);
            boolean result = nextBuild != null && (
                //same size
                (nextBuild.block.size == size && dest.tileX() + Geometry.d4(source.rotation).x * size == nextBuild.tileX() && dest.tileY() + Geometry.d4(source.rotation).y * size == nextBuild.tileY()) ||

                    //differing sizes
                    (nextBuild.block.size > size &&
                        (source.rotation % 2 == 0 ? //check orientation
                            Math.abs(nextBuild.y - dest.y) <= (nextBuild.block.size * tilesize - size * tilesize) / 2f : //check Y alignment
                            Math.abs(nextBuild.x - dest.x) <= (nextBuild.block.size * tilesize - size * tilesize) / 2f   //check X alignment
                        )));
            if(result && nextBuild.block.outputsPayload && !nextBuild.tile.solid() && (nextBuild.rotation + 2) % 4 != source.rotation){
                result = nextBuild.acceptPayload(source, payload);
                if(result) nextBuild.handlePayload(source, payload);
                return result;
            }else return false;
        }
    }
}
